### PR TITLE
Make all changes easier to review

### PR DESCRIPTION
**Pull request by trieloff**
See #13 @tripodsan

### CODE REVIEW

**Comment by tripodsan** [comment]
thanks....but I think it's the wrong way around :-) 

you could...
- branch first
- reset master
- rebase on master
- create PR....

**Comment by trieloff** [comment]
I don't want to reset the master. This PR can be reviewed easily, and it keeps the history.

**Comment by tripodsan** [comment]
> I don't want to reset the master. This PR can be reviewed easily, and it keeps the history.

so, you will update this branch and then eventually merge back to master?

**Comment by tripodsan** [comment]
anyways, @trieloff, if you think it's ready, let's ask the team to review.

**Comment by rofe** [comment]
My apologies, but what's the purpose of this repo again? 😬

File: src/index.js [hunk 1]
```diff
@@ -13,16 +13,118 @@ const { wrap } = require('@adobe/openwhisk-action-utils');
...
+];
+
+const HANDLERS = [
+  github, google, onedrive,
+];
 
 /**
- * This is the main function
- * @param {string} name name of the person to greet
+ * Retrives content from a content repository
```
Comment (stefan-guggisberg) [inline src/index.js:35]: Retrieves

Comment (trieloff) [inline src/index.js:35]:
```suggestion
 * Retrieves content from a content repository
```

File: src/cache.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ * @param {function} cacheresult a predicate function that will be called
+ * with the result of the function execution. Only when this function returns
+ * `true` will the result be cached. The default is that all results are
+ * cached.
+ * @param {function} cacheerror a predicate function that will be called
+ * with the error that the function execution throws. Only when this
+ * function returns `true` will the error be cached. The default is that
+ * errors are never cached, so that each new invocation will be uncached.
+ */
+function cache(fn, {
```
Comment (tripodsan) [inline src/cache.js:33]: Change to explicit `options` argument that allows better documentation:

```js
/**
 * Returns a memoized version of the function `fn`.
 * @param {function} fn a function to memoize
 * @param {CacheOptions} opts options to configure the cache
 * @param {function} [opts.hash] a hash function to build a cache key.
 *          The hash function will be called with the `fn` and a list of
 *          its arguments. A good hashing function will discard irrelevant
 *          arguments and join the resulting arguments
 * @param {function} [opts.cacheresult] a predicate function that will be called
 *          with the result of the function execution. Only when this function returns
 *          `true` will the result be cached. The default is that all results are
 *          cached.
 * @param {function} [opts.cacheerror] a predicate function that will be called
 *          with the error that the function execution throws. Only when this
 *          function returns `true` will the error be cached. The default is that
 *          errors are never cached, so that each new invocation will be uncached.
 */
function cache(fn, opts = {}) {
  const {
     hash = (...args) => args.join(),
    cacheresult = () => true,
    cacheerror = () => false,
  } = opts;
  .
  .
}
```

File: src/cache.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        return result.ok;
+      }
+      throw result.err;
+    }
+    try {
+      // invoke the function
+      const result = await fn(...args);
+      if (cacheresult(result)) {
+        // store the result under ok if permitted
+        lru.set(key, { ok: result });
```
Comment (tripodsan) [inline src/cache.js:53]: edge case: what if `result` is _falsy_? then the test above would not work.
what about storing the `result` or `error` directly as value, and use `instanceof Error` to check if it was an error?

File: src/cache.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        lru.set(key, { err });
+      }
+      // and throw the error
+      throw err;
+    }
+  };
+}
+
+cache.options = (opts) => {
+  lru = new QuickLRU(opts);
```
Comment (tripodsan) [inline src/cache.js:69]: careful: every action that invokes `cache.options` will destroy the cache. 
what about creating the cache the first time it is needed?

File: src/github.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+const URL = require('url');
+const { fetch } = require('@adobe/helix-fetch').context({
+  httpsProtocols:
+  /* istanbul ignore next */
+  process.env.FORCE_HTTP1 ? ['http1'] : ['http2', 'http1'],
```
Comment (tripodsan) [inline src/github.js:16]: can we call this:

```suggestion
  process.env. HELIX_FETCH_FORCE_HTTP1 ? ['http1'] : ['http2', 'http1'],
```

(https://github.com/adobe/helix-documents-support/blob/6fa02c6480089bbef03de98d3f63e68c6b103d55/src/BlobHandler.js#L89)

File: src/github.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    /\/+/g,
+    '/',
+  );
+
+  rootURI.pathname = fullPath;
+
+  return URL.format(rootURI);
+}
+
+async function fetchFSTabImpl({
```
Comment (tripodsan) [inline src/github.js:35]: add distinct options object for better documentation

File: src/github.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  root, owner, repo, ref, log, options,
+}) {
+  const response = await fetch(computeGithubURI(root, owner, repo, ref, 'fstab.yaml'), options);
+  if (response.ok) {
+    return response.text();
+  } else if (response.status === 404) {
+    log.info(`No fstab.yaml found in repo ${owner}/${repo}`);
```
Comment (tripodsan) [inline src/github.js:42]: careful: always read-out the response, otherwise fetch might hang.
better to read the body right away:

```js
const response = await fetch(computeGithubURI(root, owner, repo, ref, 'fstab.yaml'), options);
const body = await response.text();
```

File: src/github.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+  err.status = utils.propagateStatusCode(response.status);
+  throw err;
+}
+
+// keep it cachy.
+const fetchFSTab = cache(fetchFSTabImpl, {
+  hash: (fn, {
+    owner, repo, ref, _, options,
+  }) => ([
+    fn.name,
```
Comment (tripodsan) [inline src/github.js:56]: what is `fn.name` ?

File: src/github.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+  }
+  log[utils.logLevelForStatusCode(response.status)](`Invalid response (${response.status}) when fetching fstab for ${owner}/${repo}`);
+  const err = new Error('Unable to fetch fstab', await response.text());
+  err.status = utils.propagateStatusCode(response.status);
+  throw err;
+}
+
+// keep it cachy.
+const fetchFSTab = cache(fetchFSTabImpl, {
+  hash: (fn, {
```
Comment (tripodsan) [inline src/github.js:53]: can you move this to a different function? looks a bit confusing....

File: src/github.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    fn.name,
+    owner,
+    repo,
+    ref,
+    options && options.headers && options.headers.Authorization
+      ? options.headers.Authorization : undefined,
+  ].join()),
+});
+
+function isimmutable(ref) {
```
Comment (tripodsan) [inline src/github.js:65]:
```suggestion
function isImmutable(ref) {
```

File: src/github.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  return ref && ref.match(/^[a-f0-9]{40}$/i);
+}
+
+async function handle({
```
Comment (tripodsan) [inline src/github.js:69]: add distinct options...

what about adding a `HandlerOptions` declaration to a `types.d.ts` or at least as jsdoc. I should be the same for all handlers.

File: src/github.js [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+  githubRootPath, owner, repo, ref, path, log, options,
+}) {
+  const uri = computeGithubURI(githubRootPath, owner, repo, ref, path);
+  const response = await fetch(uri, options);
```
Comment (tripodsan) [inline src/github.js:73]: same as above: better to read out the `body` in anycase.

File: src/github.js [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+  if (response.ok) {
+    return {
+      statusCode: 200,
+      body: await response.text(),
+      headers: {
+        'content-type': 'text/plain',
+        'x-source-location': uri,
+        'cache-control': isimmutable(ref) ? 'max-age=30758400' : 'max-age=60',
+        'surrogate-control': isimmutable(ref) ? 'max-age=30758400, stale-while-revalidate=30758400, stale-if-error=30758400, immutable' : 'max-age=60',
```
Comment (tripodsan) [inline src/github.js:82]: `isImmutable` is called twice... could be optimzied.

also,In the README it says, that content-proxy only handles immutable refs anyways....

File: src/google.js [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+const { URL } = require('url');
+
+const { fetch } = require('@adobe/helix-fetch').context({
+  httpsProtocols:
+  /* istanbul ignore next */
+  process.env.FORCE_HTTP1 ? ['http1'] : ['http2', 'http1'],
```
Comment (tripodsan) [inline src/google.js:17]: see similar comment in `github.js`

File: src/google.js [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+});
+const { utils } = require('@adobe/helix-shared');
+
+async function handle({
```
Comment (tripodsan) [inline src/google.js:21]: see similar comment in `github.js`

File: src/google.js [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+  url.searchParams.append('rootId', mp.id);
+
+  url.searchParams.append('rid', options.requestId);
+  url.searchParams.append('src', `${owner}/${repo}/${ref}`);
+
+  const fetchopts = options;
+  delete fetchopts.requestId;
+
+  const response = await fetch(url.href, fetchopts);
+  if (response.ok) {
```
Comment (tripodsan) [inline src/google.js:36]: better to always read `body`

File: src/google.js [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    return {
+      body: await response.text(),
+      statusCode: 200,
+      headers: {
+        'content-type': 'text/plain',
+        // if the backend does not provide a source location, use the URL
+        'x-source-location': await response.headers.get('x-source-location') || url.href,
```
Comment (tripodsan) [inline src/google.js:43]: doesn't need `await`

File: src/index.js [hunk 18]
```diff
@@ -13,16 +13,118 @@ const { wrap } = require('@adobe/openwhisk-action-utils');
...
+      }
+      return headers;
+    }, {
+      // pass on authorization token
+      Authorization: gitHubToken ? `token ${gitHubToken}` : undefined,
+    }),
+  };
+
+  const externalOptions = {
+    timeout: HTTP_TIMEOUT ? HTTP_TIMEOUT * 10 : 10 * 1000,
```
Comment (tripodsan) [inline src/index.js:72]: use separate timeout, as defined in:

https://github.com/adobe/helix-pipeline/blob/b8d87656c16e7494cd8432ae58c59945e847df03/src/schemas/secrets.schema.json#L73-L77

(also, default is 20s)

File: src/index.js [hunk 19]
```diff
@@ -13,16 +13,118 @@ const { wrap } = require('@adobe/openwhisk-action-utils');
...
+    || '',
   };
+
+  try {
+    const fstab = await fetchFSTab({
+      root: githubRootPath, owner, repo, ref, log, options: githubOptions,
+    });
+    const mount = await new MountConfig().withSource(fstab).init();
+
+    // black magic from helix-pipeline
```
Comment (tripodsan) [inline src/index.js:85]:
```suggestion
    // extract resource path w/o extension.
    // eg: /foo.bar/welcome.gift.md -> /foo.bar/welcome.gift
```

File: src/onedrive.js [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+const { URL } = require('url');
+
+const { fetch } = require('@adobe/helix-fetch').context({
+  httpsProtocols:
+  /* istanbul ignore next */
+  process.env.FORCE_HTTP1 ? ['http1'] : ['http2', 'http1'],
```
Comment (tripodsan) [inline src/onedrive.js:17]: see comment in github.js

File: src/onedrive.js [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+  url.searchParams.append('shareLink', mp.url);
+
+  url.searchParams.append('rid', options.requestId);
+  url.searchParams.append('src', `${owner}/${repo}/${ref}`);
+
+  const fetchopts = options;
+  delete fetchopts.requestId;
+
+  const response = await fetch(url.href, fetchopts);
+  if (response.ok) {
```
Comment (tripodsan) [inline src/onedrive.js:35]: better to read the `body` right  away

File: src/onedrive.js [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+    return {
+      body: await response.text(),
+      statusCode: 200,
+      headers: {
+        'content-type': 'text/plain',
+        // if the backend does not provide a source location, use the URL
+        'x-source-location': await response.headers.get('x-source-location'),
```
Comment (tripodsan) [inline src/onedrive.js:42]: `await` not needed.

File: src/onedrive.js [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+const { URL } = require('url');
+
+const { fetch } = require('@adobe/helix-fetch').context({
+  httpsProtocols:
+  /* istanbul ignore next */
+  process.env.FORCE_HTTP1 ? ['http1'] : ['http2', 'http1'],
+});
+const { utils } = require('@adobe/helix-shared');
+
+async function handle({
```
Comment (tripodsan) [inline src/onedrive.js:21]: see comment in github.js

File: test/github.test.js [hunk 24]
```diff
@@ -0,0 +1,290 @@
...
+const index = require('../src/index.js').main;
+const { setupPolly } = require('./utils.js');
+
+describe('GitHub Integration Tests', () => {
+  setupPolly({
+    recordIfMissing: false,
+  });
+
+  after(() => {
+    delete process.env.FORCE_HTTP1;
```
Comment (tripodsan) [inline test/github.test.js:27]: not needed.... since the fetch context is already initialized globally.

File: test/google.test.js [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+  /g: https://drive.google.com/drive/u/0/folders/1bH7_28a1-Q3QEEvFhT9eTmR-D7_9F4xP
+`;
+
+describe('Google Integration Tests', () => {
+  setupPolly({
+    recordIfMissing: false,
+  });
+
+  after(() => {
+    delete process.env.FORCE_HTTP1;
```
Comment (tripodsan) [inline test/google.test.js:33]: not needed.... since the fetch context is already initialized globally.

File: test/google.test.js [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+/* eslint-env mocha */
+
+process.env.FORCE_HTTP1 = 'true';
+
+const assert = require('assert');
+const index = require('../src/index.js').main;
```
Comment (tripodsan) [inline test/google.test.js:18]:
```suggestion
const { main } = require('../src/index.js');
```

File: test/index.test.js [hunk 27]
```diff
@@ -16,15 +16,60 @@
 
 const assert = require('assert');
 const index = require('../src/index.js').main;
```
Comment (tripodsan) [inline test/index.test.js:18]:
```suggestion
const { main } = require('../src/index.js');
```

File: test/index.test.js [hunk 28]
```diff
@@ -16,15 +16,60 @@
...
 
 describe('Index Tests', () => {
+  setupPolly({
+    recordIfMissing: false,
+  });
+
   it('index function is present', async () => {
     const result = await index({});
-    assert.deepEqual(result, { body: 'Hello, world.' });
+    assert.deepEqual(result.statusCode, 400);
```
Comment (tripodsan) [inline test/index.test.js:28]:
```suggestion
    assert.equal(result.statusCode, 400);
```

File: test/index.test.js [hunk 29]
```diff
@@ -16,15 +16,60 @@
```
Comment (tripodsan) [inline test/index.test.js:16]: don't you need the `process.env.FORCE_HTTP1` ?

File: test/onedrive.test.js [hunk 30]
```diff
@@ -0,0 +1,74 @@
...
+  /: https://adobe.sharepoint.com/sites/TheBlog/Shared%20Documents/theblog
+`;
+
+describe('OneDrive Integration Tests', () => {
+  setupPolly({
+    recordIfMissing: false,
+  });
+
+  after(() => {
+    delete process.env.FORCE_HTTP1;
```
Comment (tripodsan) [inline test/onedrive.test.js:32]: not needed.... since the fetch context is already initialized globally.

File: src/github.js (hunk 8 above)
Comment (trieloff) [reply src/github.js:56]: The name of the function. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/name

File: test/index.test.js (hunk 29 above)
Comment (trieloff) [reply test/index.test.js:16]: Not here.

File: src/cache.js (hunk 4 above)
Comment (trieloff) [reply src/cache.js:77]: I've documented the current behavior.
